resolver-dns/src/main/java/io/netty/resolver/dns/DnsAddressResolveContext.java

```diff
@@ -61,6 +62,31 @@ boolean containsExpectedResult(List<InetAddress> finalResult) {
         return false;
     }
 
+    @Override
+    List<InetAddress> filterResults(List<InetAddress> unfiltered) {
+        final Class<? extends InetAddress> inetAddressType = parent.preferredAddressType().addressType();
+        final int size = unfiltered.size();
+        int numExpected = 0;
+        for (int i = 0; i < size; i++) {
+            InetAddress address = unfiltered.get(i);
+            if (inetAddressType.isInstance(address)) {
+                numExpected++;
+            }
+        }
+        if (numExpected == size || numExpected == 0) {
+            // If all the results are the preferred type, or none of them are, then we don't need to do any filtering.
+            return unfiltered;
+        }
+        List<InetAddress> filtered = new ArrayList<InetAddress>(numExpected);
```

Comment:
@anuraaga can you just lazy create the ArrayList if needed and so remove the need to iterate two times ?

Reply:
I wanted to avoid creating the array at all if there is no need to filter, which is pretty common. I can't see a way to do that with just one loop since all the records need to be confirmed before knowing whether it needs filtering. I could merge this into `containsExpectedResult` by changing it to return `null` when it currently returns `false` or a filtered list otherwise which cuts from three to two loops. What do you think?

Reply:
actually after thinking more about it I think its ok... let me merge it